File: java/src/model/file/OpenFileStmt.java
package model.file;
import java.io.*;
import java.util.Map;

import model.ProgState;
import model.statement.Statement;
import utils.*;

public class OpenFileStmt implements Statement{
	private String fileName;
	private String idDescName;
	
	public OpenFileStmt(String name,String id){
		fileName = name;
		idDescName = id;
	}
	
	private boolean isOpen(String fileName2, IFileTable<Integer,FileData> fileTable) {
		for(Map.Entry<Integer, FileData> it: fileTable.getAll()){
			if(fileName2.equals(it.getValue().getFileName()))
				return true;
		}
		return false;
	}
	
	@Override
	public ProgState execute(ProgState p){
		if(isOpen(fileName,p.getFileTable()))
			throw new InterpretorException("file has already been opened");
		try{
			BufferedReader buffer = new BufferedReader(new FileReader(fileName));
			FileData fd = new FileData(fileName,buffer);
			int id = IdGenerator.generator();
			p.getFileTable().add(id,fd);
			p.getSymbolTable().add(idDescName, id);
		}
		catch(IOException e){
			throw new InterpretorException("cannot open file",e);
		}
		return null;
	} 
	
	public String toString(){
		StringBuffer s=new StringBuffer();
		s.append("Open: ");
		s.append(fileName);
		return s.toString();
	}

}

File: C#/ToyLanguageC/model/OpenFileStmt.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using ToyLanguageC.utils;

namespace ToyLanguageC.model
{
    class OpenFileStmt : Statement
    {
        private string fileId;
        private string fileName;

        public OpenFileStmt(string fileId, string fileName)
        {
            this.fileId = fileId;
            this.fileName = fileName;
        }

        private bool isOpen(String fileName2, FileTable<int, FileData> fileTable)
        {
            foreach (KeyValuePair<int, FileData> it in fileTable.getAll())
            {
                if (it.Value.Name==fileName2)
                    return true;
            }
            return false;
        }

        public ProgState execute(ProgState state)
        {
            FileTable<int, FileData> fileT = state.getFileTable;
            if (isOpen(fileName, fileT))
                throw new InterpreterException("error: file could not be open");
            try
            {
                StreamReader bufferedReader = new StreamReader(fileName);
                FileData fd = new FileData(fileName, bufferedReader);
                int id = Generator.generator();
                SymbolTable<string, int> symT = state.getSymbolTable;
                fileT.add(id, fd);
                symT.add(fileId, id);
            }
            catch (IOException e)
            {
                throw new InterpreterException("error: file could not be open");
            }
            return state;
        }

        public string toString()
        {
            return "openRFile(" + fileId + ", \"" + fileName + "\")";
        }
    }
}
